windows/src/developer/kmcmpdll/UnreachableRules.cpp

```diff
@@ -0,0 +1,48 @@
+
+#include "pch.h"
+
+#include <compfile.h>
+#include <compiler.h>
+#include <comperr.h>
+#include <vkeys.h>
+#include <kmcmpdll.h>
+
+#include <unordered_map>
+#include <unordered_set>
+#include <string>
+#include <sstream>
+
+#include "UnreachableRules.h"
+
+std::wstring MakeHashKeyFromFileKey(PFILE_KEY kp) {
+  std::wstringstream key;
+  key << kp->Key << "," << kp->ShiftFlags << ",";
+  if (kp->dpContext) key << kp->dpContext;
+  return key.str();
+}
+
+DWORD VerifyUnreachableRules(PFILE_GROUP gp) {
+  PFILE_KEY kp = gp->dpKeyArray;
+  DWORD i;
+
+  std::unordered_map<std::wstring, FILE_KEY> map;
+  std::unordered_set<int> reportedLines;
+
+  for (i = 0; i < gp->cxKeyArray; i++, kp++) {
+    std::wstring key = MakeHashKeyFromFileKey(kp);
+    if (map.count(key) > 0) {
+      FILE_KEY const & k1 = map.at(key);
+      if (kp->Line != k1.Line && reportedLines.count(kp->Line) == 0) {
+        reportedLines.insert(kp->Line);
+        currentLine = kp->Line;
+        wsprintf(ErrExtra, "Overridden by rule on line %d", k1.Line);
+        AddWarning(CINFO_UnreachableRule);
```

Comment:
Do you want to add this?
```
return CERR_INFO;
```

Currently, it looks like `VerifyUnreachableRules()` always returns `CERR_None`

Reply:
`CERR_INFO` is a bitmask value that is not valid on its own.

Returning `CERR_None` is correct. Returning a value other than `CERR_None` aborts the compile, and suggests that the `VerifyUnreachableRules()` function had a failure. Given it only ever tags rules with info values (despite `AddWarning` it's actually a info/hint), we don't want to abort the compile.